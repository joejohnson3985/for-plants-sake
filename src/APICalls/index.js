import { hpapikey, gmapikey } from '../apikey.js';


export const getTrails = filters => {
  const { lat, lng, maxDistance, minStars, maxResults, sort, minLength } = filters
  const base = 'https://www.hikingproject.com/data/get-trails?'
  const url = `${base}lat=${lat}&lon=${lng}&maxDistance=${maxDistance}&key=${hpapikey}&minStars=${minStars}&maxResults=${maxResults}&sort=${sort}&minLength=${minLength}`
  return fetch(url).then(response => {
    if (!response.ok) {
      throw new Error('Unable to fetch trails.');
    }
    return response.json();
  })
};

export const getSearchedTrails = address => {
  const url = `https://maps.googleapis.com/maps/api/geocode/json?address=${address}&key=${gmapikey}`
  return fetch(url).then(response => {
    if (!response.ok) {
      throw new Error('Error grabbing that location.');
    }
    return response.json();
  })
}

export const getCurrentLocationName = location => {
  const { lat, lng } = location
  const url = `https://maps.googleapis.com/maps/api/geocode/json?latlng=${lat},${lng}&key=${gmapikey}`
  return fetch(url).then(response => {
    if (!response.ok) {
      throw new Error('Unable to name this location.');
    }
    return response.json();
  })
}

export const getMyHikes = (ids) => {
  const base = 'https://www.hikingproject.com/data/get-trails-by-id?'
  const url = `${base}ids=${ids}&key=${hpapikey}`
  return fetch(url).then(response => {
    if(!response.ok) {
      throw new Error('Error fetching your hikes.')
    }
    return response.json()
  })
}  

