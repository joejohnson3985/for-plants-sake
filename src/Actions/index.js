export const setTrails = trails => ({
  type: 'SET_TRAILS',
  trails
})