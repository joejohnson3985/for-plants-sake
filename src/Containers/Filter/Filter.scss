@import '../../Utilities/_Variables.scss';

.filter {
  display: flex;
  align-items: center;
  justify-content: space-around;
  height: 10vh;
  margin: 0 100px;
  background-color: rgba(255,255,255, 0.6);


  .filter-options {
    margin: 0 10px;
    font-size: 18px;
    color: $dark-text;
  }

  .submit-btn {}
}