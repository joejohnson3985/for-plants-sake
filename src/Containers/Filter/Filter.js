import React, { Component } from 'react';
import './Filter.scss'
import { getTrails, getSearchedTrails, getCurrentLocationName } from '../../APICalls'
import { setTrails } from '../../Actions';
import { connect } from 'react-redux';

class Filter extends Component {
  constructor() {
    super()
    this.state = {
      search:'',
      maxDistance: 20,
      lat: 0,
      lng: 0,
      sort: 'quality',
      minLength: 0,
      minStars: 0,
      maxResults: 20
    }
  }

  componentDidMount() {
    this.getLocation()
  }

  getLocation = () => {
    const success = this.setCurrentLocation
    const error = this.errorLocating
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(success, error )
    } else {
      console.log('not working');
    }
  }

  setCurrentLocation = (pos) => {
    const { latitude, longitude } = pos.coords
    getCurrentLocationName({lat: latitude, lng: longitude})
      .then(query => this.findLocalAdress(query.results))
      .then(address => this.setState({search: address.formatted_address, lat: latitude, lng: longitude}, () => {
        this.handleSubmit()
        }
      ))
  }

  findLocalAdress = (results) => (results.find(result => result.types.includes('locality')))

  errorLocating = (positionError) => {
    console.log(positionError)
    this.setState({lat: 39.7392358, lng: -104.990251})
    this.handleSubmit()
  }

  handleChange = (e) => {
    const { name, value } = e.target
    this.setState({[name]: value})
  }

  handleSearch = (e) => {
    e.preventDefault();
    getSearchedTrails(this.state.search)
    .then(results => {
      const { lat, lng } = results.results[0].geometry.location
      this.setState({lat, lng})

    })
  }

  handleSubmit = () => {
    getTrails(this.state)
    .then(results => this.props.setTrails(results.trails))
  }

  render() {
    const { sort, maxDistance, minStars, search, minLength} = this.state
    return(
      <form className='filter'>
        <input className='filter-options' name='search' value={search} onBlur={this.handleSearch} onChange={this.handleChange} type='text'/>
          <input className='filter-options' name='maxDistance' onChange={this.handleChange} value={maxDistance} type='number'/>
          <select className='filter-options' value={sort} name='sort' onChange={this.handleChange}>
            <option value='quality'>Quality</option>
            <option value='distance'>Distance</option>
          </select>
          <select className='filter-options' value={minStars} name='minStars' onChange={this.handleChange}>
            <option value={0}>All Ratings</option>
            <option value={1}>1</option>
            <option value={2}>2</option>
            <option value={3}>3</option>
            <option value={4}>4</option>
            <option value={5}>5</option>
          </select>
          <input className='filter-options' name='minLength' onChange={this.handleChange} value={minLength} type='number'/>
          <div className='submit-btn' role='button' onClick={this.handleSubmit}>Update Results</div>
      </form>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  setTrails: trails => dispatch(setTrails(trails)),
});

export default connect(null, mapDispatchToProps)(Filter);