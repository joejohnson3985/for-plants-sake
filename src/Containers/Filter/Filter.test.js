import React from 'react';
import { shallow } from 'enzyme';
import { Filter, mapDispatchToProps } from './Filter';
import { setTrails, setLoading, setError } from '../../Actions';
import { getCurrentLocationName, getSearchedTrails, getTrails } from '../../APICalls'

jest.mock('../../APICalls/index.js');
getCurrentLocationName.mockImplementation(() => Promise.resolve(1));
getSearchedTrails.mockImplementation(() => Promise.resolve(1));
getTrails.mockImplementation(() => Promise.resolve(1));

let wrapper;
let expected;
let result;
let mockSetLoading;
let mockSetError;
let mockPos;
let mockEvent;
let mockAddress1 = {address: 'Whittier, Denver, CO 80205, USA', types:['political']}
let mockAddress2 = {address: '2445 N Gaylord Street, Denver, CO, 80205', types:['street address']}
let mockAddresses = [mockAddress1, mockAddress2]
let mockPositionError = {message: 'mock error message'}
let mockFilter = {
        search:'Test',
        maxDistance: 20,
        lat: 0,
        lng: 0,
        sort: 'quality',
        minLength: 0,
        minStars: 0,
        maxResults: 20
      }


describe('Filter', () => {

  beforeEach(() => {
    mockSetLoading = jest.fn()
    mockSetError = jest.fn()
    wrapper= shallow(
      <Filter setLoading={mockSetLoading} setError={mockSetError} />
    )
  })

  describe('Component', () => {

    it('Should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    })

    it('should have a default state', () => {
      expected = {
        search:'',
        maxDistance: 20,
        lat: 0,
        lng: 0,
        sort: 'quality',
        minLength: 0,
        minStars: 0,
        maxResults: 20
      }
      expect(wrapper.state()).toEqual(expected)
    })
  })

  describe('Methods', () => {

    describe('getLocation', () => {
      it('Should call setLoading when when getLocation is invoked', () => {
        wrapper.instance().getLocation()
        expect(mockSetLoading).toHaveBeenCalledWith(true)
      })
    })

    describe('setCurrentLocation', () => {
      mockPos = {coords: {latitude: 39.7392358, longitude: -104.990251}}
      const { latitude, longitude } = mockPos.coords
      let expected = {lat: latitude, lng: longitude}

      it('Should call getCurrentLocationName with the correct params', () => {
        wrapper.instance().setCurrentLocation(mockPos)
        expect(getCurrentLocationName).toHaveBeenCalledWith(expected)
      })

      it('Should return an object with a type that includes political', () => {
        result = wrapper.instance().findLocalAddress(mockAddresses)
        expect(result).toEqual(mockAddress1)
      })
    })

    describe('errorLocating', () => {

      it('Should call the function setError with correct params when invoked', () => {
        let expectedError = 'mock error message. Default location is Denver, Colarado.'
        wrapper.instance().errorLocating(mockPositionError)
        expect(mockSetError).toHaveBeenCalledWith(expectedError)
      })

      it.skip('Should call the function setCurrentLocation when invoked', () => {
        mockPos = {coords: {latitude: 39.7392358, longitude: -104.990251}}
        wrapper.instance().errorLocating(mockPositionError)
        expect(wrapper.setCurrentLocation).toHaveBeenCalledWith(mockPos)
      })
    })

    describe('handleChange', () => {
      it('Should set state when invoked', () => {
        mockEvent = {target: {name: 'maxDistance', value: 30}}
        expect(wrapper.state().maxDistance).toEqual(20)
        wrapper.instance().handleChange(mockEvent)
        expect(wrapper.state().maxDistance).toEqual(30)
      })
    })

    describe('handleSearch', () => {
      it('Should call getSearchedTrails with the correct params', () => {
        wrapper.state().search = 'test search'
        wrapper.instance().handleSearch()
        expect(getSearchedTrails).toHaveBeenCalledWith('test search')
      })
    })

    describe('handleSubmit', () => {
      it('Should call getTrails with correct params', () => {
        wrapper.state().search = 'Test'
        wrapper.instance().handleSubmit()
        expect(getTrails).toHaveBeenCalledWith(mockFilter)
      })
    })

    
  })

  describe('mapDispatchToProps', () => {
    it('Should call setTrails with correct params', () => {
      let mockTrails = [{number: 1}, {number: 2}]
      const mockDispatch = jest.fn();
      const actionToDispatch = setTrails(mockTrails)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setTrails(mockTrails)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('Should call setError with correct params', () => {
      let mockError = 'error'
      const mockDispatch = jest.fn();
      const actionToDispatch = setError(mockError)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setError(mockError)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('Should call setLoading with correct params', () => {
      let mockBool = true
      const mockDispatch = jest.fn();
      const actionToDispatch = setLoading(mockBool)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setLoading(mockBool)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })  })
})