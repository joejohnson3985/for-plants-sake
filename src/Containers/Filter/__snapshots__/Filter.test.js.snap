// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Component Should match the snapshot 1`] = `
<form
  className="filter"
>
  <div
    className="form-items"
  >
    <label
      htmlFor="search"
    >
      Location
    </label>
    <input
      className="filter-options"
      name="search"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-items"
  >
    <label
      htmlFor="maxDistance"
    >
      Radius (miles)
    </label>
    <input
      className="filter-options"
      name="maxDistance"
      onChange={[Function]}
      type="number"
      value={20}
    />
  </div>
  <div
    className="form-items"
  >
    <label
      htmlFor="minStars"
    >
      Stars
    </label>
    <select
      className="filter-options"
      name="minStars"
      onChange={[Function]}
      value={0}
    >
      <option
        value={0}
      >
        All Ratings
      </option>
      <option
        value={1}
      >
        1
      </option>
      <option
        value={2}
      >
        2
      </option>
      <option
        value={3}
      >
        3
      </option>
      <option
        value={4}
      >
        4
      </option>
      <option
        value={5}
      >
        5
      </option>
    </select>
  </div>
  <div
    className="form-items"
  >
    <label
      htmlFor="minLength"
    >
      Min Trail Length (miles)
    </label>
    <input
      className="filter-options"
      name="minLength"
      onChange={[Function]}
      type="number"
      value={0}
    />
  </div>
  <div
    className="form-items"
  >
    <label
      htmlFor="sort"
    >
      Sort By
    </label>
    <select
      className="filter-options"
      name="sort"
      onChange={[Function]}
      value="quality"
    >
      <option
        value="quality"
      >
        Quality
      </option>
      <option
        value="distance"
      >
        Distance
      </option>
    </select>
  </div>
  <div
    className="submit-btn"
    onClick={[Function]}
    role="button"
  >
    Update Results
  </div>
</form>
`;
